File: MC_limiter.m
function [x] = MC_limiter(a,b,dx,max_t,dt)
        %initial conditions
        function [q_0] = q_0(x)
            if x <= 1
                q_0 = 1;
            else
                q_0 = 0;
            end
        end
        %initialize space, time, nu
        space = a:dx:b; 
        time = 0:dt:max_t;
        nu = dt/dx;
        %initialize Q
        Q=zeros(1,length(space));
        for j=1:length(space)
           Q(j) = q_0(space(j));
        end
        %initialize flux
        F=zeros(1,length(Q)+1);
        F(1) = 1;
        if Q(2)-Q(1) == 0
            F(2) = Q(1);
        else
            theta_0 = (Q(1) - 1)/(Q(2)-Q(1));
            phi_0 = max(0,min(1.5+theta_0/2,2,2*theta_0));
            F(2) = Q(1) + phi_0*0.5*(1-nu)*(Q(2)-Q(1));
        end
        for i=3:length(Q)
            if Q(i)-Q(i-1) == 0
               F(i) = Q(i-1);
            else
                theta = (Q(i-1)-Q(i-2))/(Q(i)-Q(i-1));
                phi = max(0,min([0.5+theta/2,2,2*theta]));
                F(i) = Q(i-1) + phi*0.5*(1-nu)*(Q(i)-Q(i-1));
            end
        end
        F(length(Q)+1) = Q(length(Q)); %upwind flux for rightmost edge
        %step through time
        for j=2:length(time)
            Q(1) = 1;
            for k=2:length(Q)
               Q(k) = Q(k) + (dt/dx)*(F(k)-F(k+1));
            end
            
            if Q(2)-Q(1) == 0
                F(2) = Q(1);
            else
                theta_0 = (Q(1) - 1)/(Q(2)-Q(1));
                phi_0 = max(0,min([0.5+theta_0/2,2,2*theta_0]));
                F(2) = Q(1) + phi_0*0.5*(1-nu)*(Q(2)-Q(1));
            end
            
            for i=3:length(Q)
                theta = (Q(i-1)-Q(i-2))/(Q(i)-Q(i-1));
                phi = max(0,min([0.5+theta/2,2,2*theta]));
                F(i) = Q(i-1) + phi*0.5*(1-nu)*(Q(i)-Q(i-1));
            end
            F(length(Q)+1) = Q(length(Q));
        end
    end

File: MC_limiter.cpp
//
//  MC_limiter.cpp
//  MC limiter
//
//  Created by Kevin Zhang on 12/2/18.
//  Copyright © 2018 Kevin Zhang. All rights reserved.
//

#include "MC_limiter.hpp"
#include <iostream>
#include <math.h>
#include <vector>
#include <algorithm>
#include "MC_limiter.hpp"

using namespace std;

double q_0(double x){
    double q_0;
    if (x<=1){q_0=1.0;}
    else{q_0=0.0;}
    return q_0;
}

vector<double> MC_limiter(int a, int b, double dx, double dt, int max_t){
    double nu = dt/dx;
    
    vector<double> space;
    for (int j=0; j<=(b-a)/dx; ++j){
        space.push_back(j*dx);
    }
    
    vector<double> time;
    for (int j=0; j<=max_t/dt; ++j){
        time.push_back(j*dt);
    }
    
    vector<double> Q;
    for (int j=0; j<space.size(); ++j){
        Q.push_back(q_0(space[j]));
    }
    
    vector<double> F;
    F.push_back(1);
    if (Q[1]-Q[0] == 0){
        F.push_back(Q[0]);
    }
    else {
        double theta_0 = (Q[0]-1)/(Q[1]-Q[0]);
        double phi_0 = fmax(0,fmin(0.5+theta_0/2, fmin(2,2*theta_0)));
        F.push_back(Q[0]+phi_0*0.5*(1-nu)*(Q[1]-Q[0]));
    }
    for (int i = 2; i < Q.size(); ++i){
        if (Q[i] - Q[i-1] == 0) {
            F.push_back(Q[i-1]);
        }
        else {
            double theta = (Q[i-1]-Q[i-2])/(Q[i]-Q[i-1]);
            double phi = fmax(0,fmin(0.5+theta/2, fmin(2,2*theta)));
            F.push_back(Q[i-1] + phi*0.5*(1-nu)*(Q[i]-Q[i-1]));
        }
    }
    F.push_back(*(Q.end()-1));
    
    for (int j = 1; j<time.size(); ++j) {
        Q[0] = 1;
        for (int i = 1; i < Q.size(); ++i){
            Q[i] = Q[i] + (dt/dx)*(F[i]-F[i+1]);
        }
        
        if (Q[1]-Q[0]==0){
            F[1] = Q[0];
        }
        else {
            double theta_0 = (Q[0]-1)/(Q[1]-Q[0]);
            double phi_0 = fmax(0,fmin(0.5+theta_0/2, fmin(2,2*theta_0)));
            F[1] = Q[0] + phi_0*0.5*(1-nu)*(Q[1]-Q[0]);
        }
        
        for (int i = 2; i < Q.size(); ++i){
            double theta = (Q[i-1]-Q[i-2])/(Q[i]-Q[i-1]);
            double phi = fmax(0,fmin(0.5+theta/2, fmin(2,2*theta)));
            F[i] = Q[i-1] + phi*0.5*(1-nu)*(Q[i]-Q[i-1]);
        }
        *(F.end()-1) = *(Q.end()-1);
    }
    return Q;
}
